library/core/src/num/int_format.rs

```diff
@@ -0,0 +1,256 @@
+use crate::mem::MaybeUninit;
+use crate::slice;
+
+/// A minimal buffer implementation containing elements of type
+/// `MaybeUninit<u8>`.
+#[unstable(feature = "int_format_into", issue = "138215")]
+#[derive(Debug)]
+pub struct NumBuffer<const BUF_SIZE: usize> {
+    /// An array of elements of type `MaybeUninit<u8>`.
+    ///
+    /// An alternative to `contents.len()` is `BUF_SIZE`.
+    pub contents: [MaybeUninit<u8>; BUF_SIZE],
+}
+
+#[unstable(feature = "int_format_into", issue = "138215")]
+impl<const BUF_SIZE: usize> NumBuffer<BUF_SIZE> {
+    /// Initializes `contents` as an uninitialized array of `MaybeUninit<u8>`.
+    #[unstable(feature = "int_format_into", issue = "138215")]
+    pub fn new() -> Self {
+        NumBuffer { contents: [MaybeUninit::<u8>::uninit(); BUF_SIZE] }
+    }
+}
+
+const DEC_DIGITS_LUT: &[u8; 200] = b"\
+        0001020304050607080910111213141516171819\
+        2021222324252627282930313233343536373839\
+        4041424344454647484950515253545556575859\
+        6061626364656667686970717273747576777879\
+        8081828384858687888990919293949596979899";
+
+const NEGATIVE_SIGN: &[u8; 1] = b"-";
+
+/// Writes the negative sign into the buffer.
+/// Must be called only in the case where the number is negative.
+fn raw_write_sign_into<const BUF_SIZE: usize>(
+    buf: &mut NumBuffer<BUF_SIZE>,
+    start_offset: usize,
+) -> usize {
+    let mut offset = start_offset;
+
+    // SAFETY: All of the decimals (with the sign) fit in buf, since it now is size-checked
+    // and the if condition ensures (at least) that the sign can be added.
+    unsafe { core::hint::assert_unchecked(offset >= 1) }
+
+    // SAFETY: The offset counts down from its initial value BUF_SIZE
+    // without underflow due to the previous precondition.
+    unsafe { core::hint::assert_unchecked(offset <= BUF_SIZE) }
+
+    // Setting sign for the negative number
+    offset -= 1;
+    buf.contents[offset].write(NEGATIVE_SIGN[0]);
+
+    offset
+}
+
+/// Exports the stringified version of the buffer.
+fn raw_extract_as_str<const BUF_SIZE: usize>(buf: &NumBuffer<BUF_SIZE>, offset: usize) -> &str {
+    // SAFETY: All contents of `buf` since `offset` is set.
+    let written = unsafe { buf.contents.get_unchecked(offset..) };
+
+    // SAFETY: Writes use ASCII from the lookup table
+    // (and `NEGATIVE_SIGN` in case of negative numbers) exclusively.
```

Comment:
This function should be `unsafe` since it does not *itself* ensure the properties the safety comments rely on, it relies on the caller to do so.